Description: Handle server terminating the connection
 This is a backport of the patch from STR#4187, plus the improvement 
 from upstream git 80360a5e.
--- a/cups/request.c
+++ b/cups/request.c
@@ -1004,6 +1004,26 @@
       httpClose(cg->http);
       cg->http = NULL;
     }
+    else
+    {
+     /*
+      * Same server, see if the connection is still established...
+      */
+
+      char     ch;                     /* Connection check byte */
+      ssize_t  n;                      /* Number of bytes */
+
+      if ((n = recv(cg->http->fd, &ch, 1, MSG_PEEK | MSG_DONTWAIT)) == 0 ||
+          (n < 0 && errno != EWOULDBLOCK))
+      {
+       /*
+        * Nope, close the connection...
+        */
+
+	httpClose(cg->http);
+	cg->http = NULL;
+      }
+    }
   }
 
  /*
